types/three/src/renderers/shaders/ShaderLib.d.ts

```diff
@@ -2,6 +2,7 @@ import { IUniform } from './UniformsLib.js';
 
 export interface Shader {
     uniforms: { [uniform: string]: IUniform };
+    defines?: { [define: string]: string | number | boolean };
```

Comment:
Hmm, I'm not sure this is the best place to fix this. I think instead the first parameter of `Material.onBeforeCompile` should be updated to be more accurate.

It looks like that parameter is coming from [here](https://github.com/mrdoob/three.js/blob/dev/src/renderers/WebGLRenderer.js#L1645). The `parameters` object mostly [comes from](https://github.com/mrdoob/three.js/blob/dev/src/renderers/WebGLRenderer.js#L1603)  the [return value of `WebGLPrograms.getParameters`](https://github.com/mrdoob/three.js/blob/dev/src/renderers/webgl/WebGLPrograms.js#L177-L370) with [at least one modification](https://github.com/mrdoob/three.js/blob/dev/src/renderers/WebGLRenderer.js#L1641).

Reply:
True, should I replace `Shader` with a complete `WebGLProgramParameters` interface? Then we could also specify the return type of `WebGLPrograms.getParameters`.

Reply:
Yes, that would be great.

Reply:
Note that the argument passed into `Material.onBeforeCompile` has an additional `uniforms` property that `WebGLProgramParameters` does not define, so I imagine we would also need a `WebGLProgramParametersWithUniforms` that extends from `WebGLProgramParameters` with the addition of a `uniforms` property.

Reply:
Done. I've also specified the remaining types in `WebGLPrograms`. I moved the type of `WebGLLights.state` into a separate interface. Let me know if that's ok or if I should revert it.